feat(aio): add "suggest edits" feature to all docs

martinsik opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

The "Suggest Edits" pencil icon is only in the API subsection.

Issue Number: #23981

## What is the new behavior?

The "Suggest Edits" pencil icon is on all pages under `/docs`.

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

martinsik committed:
feat(docs-infra): add "suggest edits" feature to all docs

martinsik committed:
fix(docs-infra): show "suggest edits" only for /guide and /tutorial dirs

gkalpak commented on aio/tools/transforms/templates/githubLinks.html:
```diff
@@ -1,12 +1,16 @@
+https://github.com/{$ versionInfo.gitRepoInfo.owner $}/{$ versionInfo.gitRepoInfo.repo $}/edit/master{$ '/packages' if doc.docType !== 'content' $}/{$ projectRelativePath(doc.fileInfo) $}?message=docs(
   {%- if doc.moduleDoc %}{$ doc.moduleDoc.id.split('/')[0] $}
   {%- elseif doc.docType === 'module' %}{$ doc.id.split('/')[0] $}
+  {%- elseif doc.docType === 'content' %}aio
```
`aio` is not an appropriate scope (it is only for documentation of the docs-viewer app itself or the documentation tooling; and it has been renamed to `docs-infra` to make it clear).
For guides and tutorial, it should be just `docs: ...`

mary-poppins commented:
You can preview a8b1b67 at https://pr24378-a8b1b67.ngbuilds.io/.

gkalpak reviewed (commented):
Please change the `scope` in commit messages to `docs-infra`.
Other than that, it looks good from a quick look 💯

gkalpak commented:
One more thing: I would like the button style to be closer to #24000 :grin:

petebacondarwin commented:
One more, more thing: I would rather keep helper template files (i.e. ones that are not used directly as templates) such as the githubLinks.html file out of the main folder. I understand you moved it because it is shared outside the API stuff now. Perhaps you could create a `lib` folder directly below the `templates` folder for this file?

martinsik committed:
refactor(docs-infra): refactor templates

martinsik commented:
@gkalpak I updated commit messages. The `aio` scope is actually mentioned here https://github.com/angular/angular/blob/master/CONTRIBUTING.md#scope. Now the links use just `docs: ` prefix. My changes will probably conflict with #24000 but I think visually the buttons look the same if that's what you meant.

@petebacondarwin I moved the `githubLinks.html` file to `lib` dir.

There was one very confusing thing. The `doc.fileInfo` template variable has different structure for `doc.docType === 'content'` and for everything else.

With `content` its structure comes from here https://github.com/angular/dgeni-packages/blob/8c856f3d75bb527248f86e6c70d3c6ed81e7740c/base/processors/read-files.js#L70 while all the other page types use this https://github.com/angular/dgeni-packages/blob/c7d1b54c808931394912fcf062df25aeeb572883/typescript/src/services/TsParser/FileInfo.ts. The first one has `projectRelativePath` while the second one `realProjectRelativePath` so that's why I had to add an extra condition for this.

martinsik committed:
test(docs-infra): test that the "suggest edit" buttons are visible where expected

mary-poppins commented:
You can preview 727934c at https://pr24378-727934c.ngbuilds.io/.

mary-poppins commented:
You can preview 3d093b2 at https://pr24378-3d093b2.ngbuilds.io/.

lacolaco commented on aio/tests/e2e/app.e2e-spec.ts:
```diff
@@ -179,4 +179,25 @@ describe('site App', function() {
+      expect(page.ghLinks.count()).toEqual(1);
+      /* tslint:disable:max-line-length */
+      expect(page.ghLinks.get(0).getAttribute('href'))
+        .toMatch(/https:\/\/github\.com\/angular\/angular\/edit\/master\/aio\/content\/tutorial\/toh-pt1\.md\?message=docs%3A%20describe%20your%20change\.\.\./);
```
If this feature will be out, some angular.io localization sites (at least our angular.jp) will customize it to jump to its own repository.
So I want this specs not to depend on `angular/angular` repository.

lacolaco replied:
Exactly. After I've send the above comment, I found it's e2e testing so it doesn't affect to build aio. Sorry, and thank you. 🙌

gkalpak commented:
> The `aio` scope is actually mentioned here https://github.com/angular/angular/blob/master/CONTRIBUTING.md#scope. 

Ooops, we missed it in #24295. It should be `docs-infra`. (Submitted #24410.)

> My changes will probably conflict with #24000 but I think visually the buttons look the same if that's what you meant.

AFAICT, the buttons here are blue (gray on hover), while the buttons on #24000 are gray (blue on hover).

martinsik commented:
@gkalpak When I check the latest #24000 build at https://pr24000-9697d53.ngbuilds.io/ I see blue buttons (and grey on hover).

![screen shot 2018-06-11 at 12 03 07](https://user-images.githubusercontent.com/238765/41231271-d7bb732e-6d82-11e8-8285-0cf48f8b332d.png)

gkalpak commented:
Nevermind. I was referring to the buttons for methods. I didn't realize/remember we have differently styled buttons for the whole doc. This is a different discussion, so the buttons are fine as you have them (for this PR's purpose). 👍

petebacondarwin reviewed (changes requested):
LGTM - please rebase and fix the conflicts.

martinsik commented:
@petebacondarwin rebased and conflicts fixed.

mary-poppins commented:
You can preview 2ae87a0 at https://pr24378-2ae87a0.ngbuilds.io/.

petebacondarwin commented on aio/tools/transforms/templates/content.template.html:
```diff
@@ -1,9 +1,12 @@
-<div class="page-actions">
-  <a href="{$ github.githubEditHref(doc, versionInfo) $}" aria-label="Suggest Edits" title="Suggest Edits"><i class="material-icons" aria-hidden="true" role="img">mode_edit</i></a>
-</div>
+{% if 'guide/' in relativePath or 'tutorial/' in relativePath or 'docs.md' in relativePath %}
```
What is the motivation for this change? Other than the marketing docs, which docs should not have this link?

martinsik replied:
@petebacondarwin It's been a while since I did this but I think this is because the edit icon should be visible almost everywhere including pages in the top level navigation such as `/docs` but not at `/features` or `/resources`

petebacondarwin commented:
Sorry - another rebase required ... :-(

martinsik commented:
@petebacondarwin rebased and conflicts fixed.

mary-poppins commented:
You can preview 5bc623c at https://pr24378-5bc623c.ngbuilds.io/.

mary-poppins commented:
You can preview 18d047f at https://pr24378-18d047f.ngbuilds.io/.

petebacondarwin reviewed (approved):
OK I think we are ready to go!
Thanks for your work on this @martinsik

petebacondarwin commented:
Caretaker: it appears that the rebase caused this PR's CLA status to go red. It looks to me as though it should be green.

gkalpak commented:
:tada:
